lib/ex_doc/autolink.ex

```diff
@@ -0,0 +1,397 @@
+defmodule ExDoc.Autolink do
+  @moduledoc false
+
+  # * `:app` - the app that the docs are being generated for. When linking modules they are
+  #   checked if they are part of the app and based on that the links are relative or absolute.
+  #
+  # * `:current_module` - the module that the docs are being generated for. Used to link local
+  #   calls and see if remote calls are in the same app.
+  #
+  # * `:module_id` - id of the module being documented (e.g.: `"String"`)
+  #
+  # * `:id` - id of the thing being documented (e.g.: `"String.upcase/2"`, `"library-guidelines"`, etc)
+  #
+  # * `:ext` - the extension (`".html"`, "`.xhtml"`, etc)
+  #
+  # * `:skip_undefined_reference_warnings_on` - list of modules to skip the warning on
+  defstruct [
+    :app,
+    :current_module,
+    :module_id,
+    :id,
+    ext: ".html",
+    skip_undefined_reference_warnings_on: []
+  ]
+
+  alias ExDoc.Formatter.HTML.Templates, as: T
+  alias ExDoc.Refs
+
+  @hexdocs "https://hexdocs.pm/"
+  @otpdocs "http://www.erlang.org/doc/man/"
+
+  def doc(ast, options \\ []) do
+    config = struct!(__MODULE__, options)
+    walk(ast, config)
+  end
+
+  defp walk(list, config) when is_list(list) do
+    Enum.map(list, &walk(&1, config))
+  end
+
+  defp walk(binary, _) when is_binary(binary) do
+    binary
+  end
+
+  defp walk({:pre, _, _} = ast, _config) do
+    ast
+  end
+
+  defp walk({:a, attrs, _} = ast, config) do
+    if url = custom_link(attrs, config) do
+      {:a, Keyword.put(attrs, :href, url), ast}
+    else
+      ast
+    end
+  end
+
+  defp walk({:code, _, [text]} = ast, config) do
```

Comment:
What if we have some `code` inside a `a`? Are we sure we are not putting a link inside a link?

Reply:
yup, we have a test for this: https://github.com/elixir-lang/ex_doc/pull/1132/files#diff-9ca306915521eea50659a07f823dd25fR165